### DISCUSSION TITLE
Confused about pulumi

### DISCUSSION THREAD

**Discussion by kimausloos** [pulumi/pulumi#7905] [Pulumi in Practice] [answered]
Hi,

I wanted to test pulumi a bit because it seems an interesting approach, but I'm confused a bit. I created a very small python deployment example that simply fetches a docker image and runs it with a custom command: sleep 60.

Of course the container stays up for only a minute, then it exits as I would expect. But when I run pulumi up again, the container is not started again anymore. I also deleted the stopped container and then tried again, but no success.

Isn't pulumi operating as a desired state solution? I'm asking for a running container in my code, so one should be started if it's not running.

The very small snippet I'm trying:
```
import pulumi
import pulumi_docker as docker

rocky_remote_image = docker.RemoteImage("rockylinuxRemoteImage", name="rockylinux/rockylinux")
rocky_container = docker.Container("rockyContainer", image=rocky_remote_image.latest, attach=False, command=["/usr/bin/sleep", "60"], must_run=True, rm=True)
```

**Comment by sam-cogan** [comment]
Pulumi is doing what it is meant to. Your code isn’t asking for a running container, it’s telling docker to “run this container”, there’s a difference. Pulumi is correctly configuring docker to run the container, and the container runs. The fact that the container only runs for 60 seconds is not something Pulumi cares about (based on how your code is written), it is done it’s job. When you run it again it sees that docker is still configured to run that container. 

Now, if you wanted that container to restart when it has finished, then you can use your Pulumi code to configure docker to do that. Pulumi is managing your infrastructure configuration, in this case the configuration of docker. Docker is still the one in charge of running your container.

**Reply by kimausloos** [reply]
I would agree with you, but I am asking that the container must be running, as I assume that what the 'must_run' option is for? I do admit it's just a guess as there is no explanation about the flag in the pulumi docs.

It's even weirder. When I prune everything (docker system prune) and there are no containers or images present anymore, doing `pulumi up` won't fetch the image, nor run the container. Only when I delete the pulumi state it will (once) fetch the image and start the container. So for me this is very strange behavior and seems to be counter-intuitive with the desired state principle.

When I port the example to terraform it behaves as I would expect, more specifically: if I do a terraform apply when the container is still running nothing happens. When I run an apply after 60 seconds(=container is gone) the container is created again and will run again for the requested 60 seconds.

I'm trying to run a container that will execute a command that does calculations and generates a result, exits, and will run again whenever I ask it to, so for my use case a restart policy is not something I can use.

The terraform code for reference"
```terraform
terraform {
  required_providers {
    docker = {
      source  = "kreuzwerker/docker"
      version = "2.15.0"
    }
  }
}

provider "docker" {
  host = "unix:///var/run/docker.sock"
}

resource "docker_image" "rockylinux" {
  name = "rockylinux/rockylinux:latest"
}

resource "docker_container" "rocky_container" {
  name = "rocky_container"
  image = docker_image.rockylinux.latest
  attach = false
  command = ["/usr/bin/sleep", "60"]
  must_run = true
  rm = true
}
```

**Comment by sam-cogan** [comment, marked as answer]
The Pulumi docker provider is actually based on the terraform provider so it should have the same functionality. The docs for the Pulumi container are here, which describe all of the values including must_run: https://www.pulumi.com/docs/reference/pkg/docker/container/ you can see the values in the input section. 

That said, I would agree that the must_run value should cause it to run again. I’m wondering if the rm=true option is at fault here, given the description:

If true, then the container will be automatically removed after his execution. Terraform won’t check this container after creation. Defaults to false.

This implies that the container won’t be checked after first execution.

**Reply by kimausloos** [reply]
Strange, I see the documentation is now filled in. When I was looking at it 2 days ago it was still completely empty.

About the rm=true: I'd say it should only clean up the container when it's done (docker option --rm), and that seems to match the things I see with terraform.

Anyway, I tried to remove the `rm=True` and even explicitly set it to 'False' and created a new stack on each change. Every time the container is being created but after a minute when the container stops due to the sleep being done, it still doesn't create it again.

Maybe the problems I'm having are due to an issue and I should report it?

**Comment by nimbinatus** [comment]
You are correct that must_run should start the container if it finds it stopped. I've opened a bug report from this discussion at https://github.com/pulumi/pulumi/issues/9121
